Refactor make declaration from usage code action

jmbockhorst opened the pull request:
Should fix #377 if it wasn't already fixed.

Razzeee commented on src/providers/codeAction/makeDeclarationFromUsageCodeAction.ts:
```diff
@@ -0,0 +1,82 @@
+
+    return [];
+  },
+  getFixAllCodeAction: (params: ICodeActionParams): CodeAction | undefined => {
```
Just a heads up, I've seen this feature have a bug in the sense of it being offered, when it shouldn't be. Will have to track that down somewhen, but having a fix all might make it more likely to run into it.

jmbockhorst replied:
I noticed a function that got created with the wrong number of args. It was something like:

```elm
func : Model -> String -> Model
fund arg1 =
    ...
```

We should probably use the inferred annotation as the source to determine the number of args, as it will always be correct.

Razzeee commented on src/util/utils.ts:
```diff
@@ -50,4 +52,46 @@ export class Utils {
+    return result;
+  }
+
+  public static deduplicateDiagnostics(
```
This and the next one seem like the should be moved to diagnostics.

Razzeee commented:
Do you know, why exactly we need to deduplicate the diagnostics, is that by design?

jmbockhorst commented:
It was but I just thought about it more and we shouldn’t need to.

jmbockhorst commented on src/util/types/typeInference.ts:
```diff
@@ -1041,14 +1041,19 @@ export class InferenceScope {
+
+      if (result) {
+        type = result.type;
+        this.diagnostics.push(...result.diagnostics);
```
Just need to get rid of this change, which is where to duplicates are coming from. I thought I needed this earlier but was thinking about it wrong.

jmbockhorst commented:
Another thing to note is the fix all doesn't work for type annotation fixes yet because we use the program diagnostics to compute a fix all, but the type annotation diagnostics come from the provider, not the program/checker. That's a reason I was asking if we should move those down to the type checker.

jmbockhorst commented:
Refactored a bit more to add "suggestion" diagnostics into the type checker. This means that missing type annotation diagnostic comes from the type checker and "Fix all" work for adding them. Also split normal diagnostics into "semantic" and "syntactic" (Typescript does this) to get more control over them and to always get syntax errors first.

Razzeee commented on src/util/types/typeChecker.ts:
```diff
@@ -869,4 +935,35 @@ export function createTypeChecker(workspace: IElmWorkspace): TypeChecker {
+      workspace,
+    ).diagnostics.forEach((diagnostic) => diagnostics.add(diagnostic));
+
+    // Need to check union variants
```
Is this a todo?

jmbockhorst replied:
Nope.
